File: 백준/Gold/5430. AC/AC.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

// [10, 20, 30]이라는 문자열을 reverse하면, [03, 02, 01]이 된다. 실수 주의!

public class Main {
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();

        while (t-- > 0) {
            String commands = br.readLine();
            int n = Integer.parseInt(br.readLine());
            Deque<Integer> deque = new ArrayDeque<>();
            String strings = br.readLine();
            String arr[] = strings.substring(1, strings.length() - 1).split(",");
            for (int i = 0; i < n; i++) {
                deque.offer(Integer.parseInt(arr[i]));
            }

            boolean rvsFlag = false;
            boolean errorFlag = false;
            for (int i = 0; i < commands.length(); i++) {
                if (commands.charAt(i) == 'R') {
                    rvsFlag = !rvsFlag;
                    continue;
                }
                if (deque.isEmpty()) {
                    sb.append("error\n");
                    errorFlag = true;
                    break;
                }
                if (rvsFlag) {
                    deque.pollLast();
                    continue;
                }
                deque.pollFirst();
            }
            if (!errorFlag) {
                sb.append('[');
                List<Integer> list = deque.stream().collect(Collectors.toList());
                if (rvsFlag) Collections.reverse(list);
                sb.append(list.stream().map(x -> String.valueOf(x)).collect(Collectors.joining(",")));
                sb.append("]\n");
            }
        }
        System.out.print(sb);
    }
}

File: 백준/Gold/5430. AC/AC.kt
import java.io.*;
import java.util.*;
import kotlin.math.*;

/**
 * R: 뒤집기
 * D: 첫 번째 수를 버림 (배열이 이어있는 경우 에러)
 * 커맨드는 조합해서 사용 가능
 * 배열의 초기값과 수행할 함수가 주어졌을 때, 최종 결과를 구하기
 */

fun main() = with(System.`in`.bufferedReader()) {
    var t = readLine().toInt()
    val sb = StringBuilder()
    repeat(t) {
        val commands = readLine()
        val n = readLine().toInt()
        val deque = toDeque(readLine())
        var reverseFlag = false
        var errorFlag = false

        for (command in commands) {
            if (command == 'R') {
                reverseFlag = !reverseFlag
            } else { // D
                if (deque.isEmpty()) {
                    errorFlag = true
                    break;
                }
                if (reverseFlag) {
                    deque.pollLast()
                } else {
                    deque.pollFirst()
                }
            }
        }

        if (errorFlag) {
            sb.append("error\n")
        } else {
            sb.append(toString(deque, reverseFlag)).append('\n')
        }
    }
    print(sb)
}

fun toDeque(line: String): Deque<Int> {
    return ArrayDeque(
        line.substring(1, line.length - 1)
        .split(",")
        .filter { it.isNotEmpty() }
        .map { it.toInt() }
    )
}

fun toString(deque: Deque<Int>, reverseFlag: Boolean): String {
    if (reverseFlag) {
        return deque.reversed().joinToString(prefix = "[", postfix = "]", separator = ",")
    }
    return deque.joinToString(prefix = "[", postfix = "]", separator = ",")
}